### PR TITLE
feat(decision): add support to register api keys to proxy

### PR DESCRIPTION
**Pull request by NishantJoshi00**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR adds support for exporting hashed api keys to the incoming proxy allowing for authentication outside the application facilitating detached authentication and ratelimiting 

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

The testing is done locally by running both the application and the api key registry
1. Create Merchant
image
2. The Key is added to the registry
image
3. Create API Key
image
4. Verify if the key is added to the registry
image
5. Query the service to get details
image
This response aligns with the requirements of the partial-auth system #4802 

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x]  I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/services/authentication/decision.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            Err(error)
+        }
+        Ok(response) => {
+            router_env::info!("Decision service response: {:?}", response);
+            Ok(())
+        }
+    }
+}
+
+pub async fn add_publishable_key(
```
Comment (akshay-97) [inline crates/router/src/services/authentication/decision.rs:123]: should we combine with add_api_key and pass variant to a common function ?

File: crates/router/src/core/admin.rs [hunk 2]
```diff
@@ -281,6 +281,25 @@ pub async fn create_merchant_account(
         .await
         .to_duplicate_response(errors::ApiErrorResponse::DuplicateMerchantAccount)?;
 
+    if let Some(api_key) = merchant_account.publishable_key.as_ref() {
+        let state = state.clone();
```
Comment (akshay-97) [inline crates/router/src/core/admin.rs:285]: should we clone only necessary data here?

File: crates/router/src/services/authentication/decision.rs [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    ApiKey { merchant_id: String, key_id: String },
+    /// [`PublishableKey`] is an authentication method that uses a publishable key. This is used with [`PublishableKey`]
+    PublishableKey { merchant_id: String },
+}
+
+// # Decision Service
+//
+
+pub async fn add_api_key(
+    state: &SessionState,
```
Comment (NishantJoshi00) [inline crates/router/src/services/authentication/decision.rs:74]: @akshay-97 , we cannot pass standalone `ApiClient` as the `send_request` function takes `SessionState` as an input argument

File: crates/router/src/services/authentication/decision.rs [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        RULE_ADD_METHOD,
+        &(decision_config.base_url.clone() + DECISION_ENDPOINT),
+    );
+
+    request.set_body(RequestContent::Json(Box::new(rule)));
+    request.add_default_headers();
+
+    let response = state
+        .api_client
+        .send_request(state, request, None, false)
```
Comment (NishantJoshi00) [inline crates/router/src/services/authentication/decision.rs:108]: @akshay-97 , we cannot pass standalone `ApiClient` as the `send_request` function takes `SessionState` as an input argument

File: crates/router/src/core/admin.rs (hunk 2 above)
Comment (NishantJoshi00) [reply crates/router/src/core/admin.rs:285]: https://github.com/juspay/hyperswitch/pull/5168/files#r1661123714

**Review by akshay-97 — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by akshay-97 — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

File: crates/router/src/services/authentication/decision.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    state: &SessionState,
+    api_key: Secret<String>,
+    merchant_id: String,
+    key_id: String,
+    expiry: Option<u64>,
+) -> CustomResult<(), ApiClientError> {
+    let decision_config = if let Some(config) = &state.conf.decision {
+        config
+    } else {
+        return Ok(());
```
Comment (ThisIsMani) [inline crates/router/src/services/authentication/decision.rs:86]: Add a log if possible.

**Review by ThisIsMani — APPROVED** [review]

**Pull request merged**